README.md
<h1>
    <img src="assets/img/blocks.png" align="left" width="200" style="margin-right: 20px"/>
    rpc-fetching-blocks-healthcheck
</h1> 

Simple controller written in python that serves an healthcheck to determine if an rpc node is synced and fetching blocks.

## Run 

### Locally

Run locally:

```bash
pip install -r requirements.txt
make start
```

![usage](assets/img/terminal.png)

### Docker 

Build the image:

```bash
make build
```

Specify the rpc node to check:

```bash
export RPC_NODE=https://rpc-osmosis.blockapsis.com/
```

Run:

```bash
docker run \
    -p 8080:8080 \
    -e RPC_NODE=$(RPC_NODE) \
    -e CHECK_INTERVAL=10 \
    -e NEW_BLOCK_THRESHOLD=30 \
    --name sync-controller \
    sync-controller:$(TAG) 
```

## Usage

The controller exposes a simple REST Endpoint that:

- Responds with a `200` status code if the rpc is fetching blocks
- Responds with a `503` status code if the rpc is not fetching blocks

Example:

```bash
curl -X GET --head http://0.0.0.0:8080 
HTTP/1.1 200 OK
```

## Alarms

The healthcheck can be configured to send alarms via slack webhooks:

```bash
export SLACK_WEBHOOK="https://hooks.slack.com/services/T..."
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.0.1] - 2021-04-26

### Added
- Add initial code
- Add CHANGELOG to document changes
- Add README with usage

controllers/common.py
import grequests
import requests
from collections import namedtuple

Block = namedtuple('Block', 'height time hash')

# Networking

def call_endpoint(url, headers = {'Accept': 'application/json'}):
        
    response = requests.get(url, headers=headers)
    return response

def call_endpoints(urls, headers = {'Accept': 'application/json'}):
    
    requests = (grequests.get(url, headers=headers) for url in urls)
    responses = grequests.map(requests)
    return responses

# Parsing

def parse_sync_info(response):

    catching_up = response["result"]["sync_info"]["catching_up"]
    height = response["result"]["sync_info"]["latest_block_height"]
    time = response["result"]["sync_info"]["latest_block_time"]
    hash = response["result"]["sync_info"]["latest_block_hash"]

    return catching_up, Block(int(height), time, hash)

controllers/time_controller.py
import urllib 
import logging
import asyncio
from datetime import datetime, timezone

from dateutil import parser
from slack_sdk.webhook import WebhookClient
from collections import namedtuple
from .common import call_endpoint, parse_sync_info

Block = namedtuple('Block', 'height time hash')

class TimeController():

    def __init__(self, rpc, check_interval = 10, new_block_threshold = 30, slack_webhook = None):
        self.rpc = rpc
        self.check_interval = int(check_interval)
        self.new_block_threshold = int(new_block_threshold)

        self.in_sync = None
        self.last_block = None
        self.cathing_up = None

        self.slack_webhook = slack_webhook
    
    def update_sync_info(self):
        url = urllib.parse.urljoin(self.rpc, "/status")
        response = call_endpoint(url)

        if response.status_code == 200:
            self.catching_up, self.last_block = parse_sync_info(response.json())
        else:
            logging.error("Error making the call to {url}".format(url=response.url))

        logging.debug(
            "{rpc} height={h} time={t} hash={hx} catching_up={c}"
            .format(
                rpc=self.rpc,
                h=self.last_block.height, 
                t=self.last_block.time,
                hx=self.last_block.hash,
                c=self.catching_up
            )
        )

    def update_sync_state(self, new_state):
        
        if self.in_sync != new_state:
            self.in_sync = new_state
            self.send_alarm()


    def send_alarm(self):

        if not self.slack_webhook:
            return None

        webhook = WebhookClient(self.slack_webhook)

        if not self.in_sync:
            response = webhook.send(
                text="text",
                blocks=[
                    {
                        "type": "section",
                        "text": {
                            "type": "plain_text",
                            "text": "🔄 Node out of sync 🔴",
                            "emoji": True
                        }
                    },
                    {
                        "type": "section",
                        "text": {
                            "type": "mrkdwn",
                            "text": "*Node*:\n{rpc}".format(rpc=self.rpc)
                        }
                    },
                    {
                        "type": "section",
                        "fields": [
                            {
                                "type": "mrkdwn",
                                "text": "*Last Block Height:*\n{height}".format(height=self.last_block.height)
                            },
                            {
                                "type": "mrkdwn",
                                "text": "*Last Block Time:*\n{time}".format(time=self.last_block.time)
                            }
                        ]
                    },
                    {
                        "type": "divider"
                    }
                ]
            )
        else:
            response = webhook.send(
                text="fallback",
                blocks=[
                    {
                        "type": "section",
                        "text": {
                            "type": "plain_text",
                            "text": "🔄 Node back in sync 🟢",
                            "emoji": True
                        }
                    },
                    {
                        "type": "section",
                        "text": {
                            "type": "mrkdwn",
                            "text": "*Node*:\n{rpc}".format(rpc=self.rpc)
                        }
                    },
                    {
                        "type": "divider"
                    }
                ]
            )
            return response.status_code


    async def loop(self):
        while True:
            self.update_sync_info()
            
            if self.catching_up:
                self.update_sync_state(False)
                logging.info("🚫 {rpc} not in sync [🏃 catching up ]".format(rpc=self.rpc))

            else:
                dt_now = datetime.now(timezone.utc)
                dt_last_block = parser.parse(self.last_block.time)
                delta_seconds = (dt_now - dt_last_block).total_seconds()

                # Update sync state
                self.update_sync_state(delta_seconds <= self.new_block_threshold)

                if self.in_sync:
                    logging.info("✅ {rpc} in sync [ 🕒 {s:.3f}(s) since block {b} ]".format(rpc=self.rpc, s=delta_seconds, b=self.last_block.height))
                else:
                    logging.error("🚫 {rpc} not in sync [ 🕒 {s:.3f}(s) since block {b} ]".format(rpc=self.rpc, s=delta_seconds, b=self.last_block.height))

            await asyncio.sleep(self.check_interval)


main.py
import os
import logging
import asyncio
from fastapi import FastAPI, Response, status, HTTPException

from controllers.time_controller import TimeController

# DEFAULT values for the parameters
DEFAULT_CHECK_INTERVAL = 10
DEFAULT_NEW_BLOCK_TIMEOUT = 30
DEFAULT_LOG_LEVEL = "INFO"

# RPC_NODE: url of the rpc nodes to check if it's fetching blocks
RPC_NODE = os.environ['RPC_NODE']

# SLACK_WEBHOOK: slack incoming webhook to send alerts
SLACK_WEBHOOK = os.getenv('SLACK_WEBHOOK')

# CHECK_INTERVAL: How often should I check
CHECK_INTERVAL = os.getenv('CHECK_INTERVAL', DEFAULT_CHECK_INTERVAL)

# NEW_BLOCK_THRESHOLD: how many seconds should I wait before reporting node as unhealth
NEW_BLOCK_THRESHOLD = os.getenv('FUNCTION', DEFAULT_NEW_BLOCK_TIMEOUT)

# LOG_LEVEL: log level verbosity {INFO, DEBUG}
LOG_LEVEL = os.getenv('LOG_LEVEL', DEFAULT_LOG_LEVEL)

LOG_LEVEL_DICT = {
    "INFO": logging.INFO,
    "DEBUG": logging.DEBUG,
}

logging.basicConfig(format='%(levelname)s: %(asctime)s - %(message)s', level=LOG_LEVEL_DICT[LOG_LEVEL])

app = FastAPI()

controller = TimeController(
    rpc = RPC_NODE,
    check_interval = CHECK_INTERVAL,
    new_block_threshold = NEW_BLOCK_THRESHOLD,
    slack_webhook = SLACK_WEBHOOK
)

@app.on_event('startup')
async def app_startup():

    logging.info("-"*60)
    logging.info("🛸 Starting controller")
    logging.info("")
    logging.info("RPC_NODE:            {}".format(RPC_NODE))
    logging.info("CHECK_INTERVAL:      {}[s]".format(CHECK_INTERVAL))
    logging.info("NEW_BLOCK_THRESHOLD: {}[s]".format(NEW_BLOCK_THRESHOLD))
    logging.info("-"*60)
    logging.info("")

    asyncio.create_task(controller.loop())

@app.get("/")
def healthcheck():
    if not controller.in_sync:
        raise HTTPException(status_code=503, detail="Node not in sync")
    return controller.in_sync